File: java/src/main/java/io/signans/sample/StreamingSTT.java
package io.signans.sample;

import java.io.StringReader;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.WebSocket;
import java.nio.ByteBuffer;
import java.nio.channels.FileChannel;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import jakarta.json.Json;
import jakarta.json.JsonObject;

public class StreamingSTT {
  public static void main(String[] args) {
    final CountDownLatch latch = new CountDownLatch(1);

    WebSocket.Listener listener = new WebSocket.Listener() {
      private StringBuilder textMessage = new StringBuilder();

      private void setLanguage(WebSocket webSocket) {
        System.out.println("Setting language.");
        String data = Json.createObjectBuilder()
            .add("command", StreamingSTTAPI.COMMAND_SET_LANGUAGE)
            .add("value", "ja")
            .build()
            .toString();
        webSocket.sendText(data, true);
      }

      private void setSamplingRate(WebSocket webSocket) {
        System.out.println("Setting sampling rate.");
        String data = Json.createObjectBuilder()
            .add("command", StreamingSTTAPI.COMMAND_SET_SAMPLING_RATE)
            .add("value", 16000)
            .build()
            .toString();
        webSocket.sendText(data, true);
      }

      private void sendAudio(WebSocket webSocket) {
        // Use a new thread to send data asynchronously.
        new Thread(() -> {
          Path file = Paths.get("ja.wav");
          try (FileChannel fc = FileChannel.open(file)) {
            // Send audio data to server, 4k per 0.1 second.
            ByteBuffer buf = ByteBuffer.allocate(4096);
            while (true) {
              buf.clear();
              if (fc.read(buf) == -1) {
                break;
              }
              buf.flip();
              while (buf.hasRemaining()) {
                webSocket.sendBinary(buf, true);
                TimeUnit.MILLISECONDS.sleep(100);
              }
            }
          } catch (Exception e) {
            e.printStackTrace();
          } finally {
            String data = Json.createObjectBuilder()
                .add("command", StreamingSTTAPI.COMMAND_END_STREAM)
                .build()
                .toString();
            webSocket.sendText(data, true);
          }
        }).start();
      }

      private void printRecognitionResult(String message) {
        JsonObject jsonObject = Json.createReader(new StringReader(message))
            .readObject();
        String status = jsonObject.getString("status");
        String text = jsonObject.getString("value");
        System.out.println(String.format("[%s] %s", status, text));
      }

      @Override
      public void onOpen(WebSocket webSocket) {
        System.out.println("Connected to streaming STT API.");
        webSocket.request(1);
        // Set language after connection established.
        this.setLanguage(webSocket);
      }

      @Override
      public CompletionStage<?> onText(WebSocket webSocket, CharSequence data, boolean last) {
        webSocket.request(1);
        this.textMessage.append(data);
        if (last) {
          String completeMessage = this.textMessage.toString();
          this.textMessage = new StringBuilder();
          JsonObject jsonObject = Json.createReader(new StringReader(completeMessage))
              .readObject();
          String messageType = jsonObject.getString("type");
          switch (messageType) {
          case StreamingSTTAPI.RESPONSE_LANGUAGE_READY:
            System.out.println("Language has been set.");
            this.setSamplingRate(webSocket);
            break;
          case StreamingSTTAPI.RESPONSE_SAMPLING_RATE_READY:
            System.out.println("Sampling rate has been set.");
            this.sendAudio(webSocket);
            break;
          case StreamingSTTAPI.RESPONSE_RECOGNITION_RESULT:
            System.out.println("Received recognition result:");
            this.printRecognitionResult(completeMessage);
            break;
          case StreamingSTTAPI.RESPONSE_RECOGNITION_ERROR:
            System.err.println("Recognition error");
            System.err.println(completeMessage);
            break;
          default:
            System.err.println("Unexpected message type: " + messageType);
          }
        }
        return null;
      }

      @Override
      public void onError(WebSocket webSocket, Throwable error) {
        System.err.println("WebSocket error occurred: " + error.getMessage());
      }

      @Override
      public CompletionStage<?> onClose(WebSocket webSocket, int statusCode, String reason) {
        System.out.println(String.format("WebSocket closed. Status code [%d]. Reason [%s].", statusCode, reason));
        latch.countDown();
        return null;
      }
    };

    try {
      JWT jwt = new JWT(API.ACCESS_KEY, API.SECRET_KEY, API.JWT_DURATION);
      String token = jwt.getToken();
      String path = "/api/v1/translate/stt-streaming";
      String tokenParam = "token=" + URLEncoder.encode("Bearer " + token, StandardCharsets.UTF_8);
      String url = API.WEBSOCKET_BASE_URL + path + "?" + tokenParam;
      HttpClient.newHttpClient()
          .newWebSocketBuilder()
          .buildAsync(URI.create(url), listener)
          .get();
      latch.await();
    } catch (Exception e) {
      e.printStackTrace();
    }
  }
}

File: node-js/src/streaming-stt.js
'use strict';

const fs = require('fs');
const WebSocket = require('ws');

const config = require('./config');
const { getJwt } = require('./auth');

const fsPromise = fs.promises;

const apiPath = '/api/v1/translate/stt-streaming';
const start = Date.now();

/**
* Command type sent from the client.
*/
const commandType = {
  setLanguage: 'SET_LANGUAGE',
  setSamplingRate: 'SET_SAMPLING_RATE',
  endStream: 'END_STREAM',
  endSession: 'END_SESSION',
};

/**
* Response types received from API endpoint.
*/
const responseType = {
  languageReady: 'LANGUAGE_READY',
  samplingRateReady: 'SAMPLING_RATE_READY',
  recognitionResult: 'RECOGNITION_RESULT',
  recognitionError: 'RECOGNITION_ERROR',
};

const speechData = {
  language: 'ja',
  samplingRate: 16000,
  audioFile: './audio/ja.wav',
};

const handleSessionMessage = (connection, message) => {
  const messageJSON = JSON.parse(message);
  switch (messageJSON.type) {
    case responseType.languageReady:
      // The language is set. Set the sampling rate.
      console.log('Language is set. Set sampling rate.');
      connection.send(JSON.stringify({
        command: commandType.setSamplingRate,
        value: speechData.samplingRate,
      }));
      break;
    case responseType.samplingRateReady:
      // The language is set. Send the audio data stream.
      console.log('Sampling rate is set. Send audio data stream.');
      connection.send(speechData.audioBuffer, (error) => {
        if (error) {
          console.error(error.message);
        }
        connection.send(JSON.stringify({
          command: commandType.endStream,
          }));
      });
      break;
    case responseType.recognitionResult:
      console.log(`Recognized transcript:${Date.now() - start} milliseconds`);
      console.log(messageJSON);
      break;
    case responseType.recognitionError:
      console.error('Recognition error:');
      console.error(messageJSON.value);
      // In case of error, we close the connection immediately.
      connection.send(JSON.stringify({
        command: commandType.endSession,
      }));
    break;
    default:
      console.log('Unexpected response type:');
      console.log(messageJSON.type);
  }
};

const main = async () => {
  const env = config.signans;
  const { accessKey, secretKey } = env.authConfig;
  const token = await getJwt(accessKey, secretKey);
  if (token) {
    const url = `wss://${env.host}${apiPath}?token=Bearer ${token}`;
    console.log(url);
    speechData.audioBuffer = await fsPromise.readFile(speechData.audioFile);
    const connection = new WebSocket(url);
    connection.on('open', () => {
      console.log('Connected to streaming STT API.');
      // Once connected, set the speech language.
      if (speechData.language) {
        connection.send(JSON.stringify({
          command: commandType.setLanguage,
          value: speechData.language,
        }));
      } else {
        connection.send(JSON.stringify({
          command: commandType.setSamplingRate,
          value: speechData.samplingRate,
        }));
      }
    });
    connection.on('message', (message) => {
      handleSessionMessage(connection, message);
    });
    connection.on('error', (error) => {
      console.error(error.message);
      connection.close();
    });
    connection.on('close', (code, message) => {
      console.log(`${code}: ${message}`);
      console.log('Connection closed.');
    });
  } else {
    console.error('JwtToken error');
  }
};

main();
